db/error_handler.h

```diff
@@ -53,15 +50,16 @@ class ErrorHandler {
                                      Status::Code code,
                                      Status::SubCode subcode);
 
-   Status SetBGError(const Status& bg_err, BackgroundErrorReason reason);
+   const Status& SetBGError(const Status& bg_err, BackgroundErrorReason reason);
 
-   Status SetBGError(const IOStatus& bg_io_err, BackgroundErrorReason reason);
+   const Status& SetBGError(const IOStatus& bg_io_err,
+                            BackgroundErrorReason reason);
 
-   Status GetBGError() { return bg_error_; }
+   const Status& GetBGError() const { return bg_error_; }
```

Comment:
Is it necessary to return ```const Status&``` from the ```Get*Error``` functions? Since its a getter, the caller would be expected to check the status.

Reply:
Actually it is important to return a & from these functions for a couple of reasons:
- Without a reference we needlessly make a copy of the member variable;
- If we do not return a reference, then the act of calling this function will cause the *_error_ to be marked as checked. This is because the Status copy constructor will clear the checked flag on the original;
- If we do not return a reference, then the returned copy must be explicitly checked somewhere.  This lead to a whole bunch of calls to permit unchecked errors when we returned a status we never cared about (and was checked implicitly as a member variable already).

Reply:
> * If we do not return a reference, then the returned copy must be explicitly checked somewhere.  This lead to a whole bunch of calls to permit unchecked errors when we returned a status we never cared about (and was checked implicitly as a member variable already).

It seems strange that the caller of ```GetBGError``` will not check the return status. That would be an obvious bug. I think we should let ```GetBGError``` return Status and fix the callers. FWIW, I didn't see any instance of ```PermitUncheckedError``` on the status returned by ```GetBGError```.

Reply:
Could be.  OTOH, is there a reason to return a copy of a member variable, rather than a const & ?  I thought the latter was always considered good programming practice...

Reply:
These Set* and Clear* seem quite different from Get* for obligation to check return status. Why do Set* and Clear* return a status anyway?

> I thought the latter was always considered good programming practice...

It mixes in the concerns of object lifetime and thread shared data, which can cause subtle bugs. Use with caution.

Reply:
I have no idea why Clear* and Set* return Status.  I agree it would be better if they did not, but I do not know feel as if I should tackle that issue.

Reply:
> This lead to a whole bunch of calls to permit unchecked errors when we returned a status we never cared about (and was checked implicitly as a member variable already).

I don't think you ever clarified whether this actually applies to GetBGError, because it seems to describe callers to SetBGError. In quick review, I don't see any callers of GetBGError that would have distinct behavior for returning Status& vs. Status. Anand and I prefer this one return Status.

Reply:
I can change the GetBGError back to Status if you really want, but would question why.  When we return a Status (not reference), we end up with the potential for making an extra ephemeral copy.  A majority of the calls to GetBGError are "Status s = GetBGError();", which means we copy the reference on assignment anyway.

I can change it back, but it seems more efficient and proper to use a const reference...